### PR TITLE
Only unchop mutate() result when needed

### LINKED ISSUE #5312: mutate_at does not retain information about ts type
I received an email from CRAN today that one of my packages was producing errors on some platforms (see results page linked below). I traced this back to what I think is a new issue in dplyr related to the ts class. As you can see in the reprex below, after applying ts in mutate_at the resulting columns are listed as type "numerice" rather than "ts". 

FYI The reprex below was produced on Windows with R 3.6.3. The issue does not occur on macOS but does seem to occur on most linux systems that CRAN uses for testing.

https://cran.r-project.org/web/checks/check_results_radiant.data.html

```
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(testthat)
#> 
#> Attaching package: 'testthat'
#> The following object is masked from 'package:dplyr':
#> 
#>     matches
dat <- mutate_at(mtcars, .vars = c("mpg", "cyl"), .funs = ts, start = c(1971, 1), frequency = 52)
expect_equal(dat$mpg, ts(mtcars$mpg, start = c(1971, 1), frequency = 52))
#> Error: dat$mpg not equal to ts(mtcars$mpg, start = c(1971, 1), frequency = 52).
#> Classes differ: numeric is not ts
sapply(dat, class)
#>       mpg       cyl      disp        hp      drat        wt      qsec        vs 
#> "numeric" "numeric" "numeric" "numeric" "numeric" "numeric" "numeric" "numeric" 
#>        am      gear      carb 
#> "numeric" "numeric" "numeric"
```

### PR DESCRIPTION
**Pull request by hadley**
Fixes #5312

### CODE REVIEW

File: R/mutate.R [hunk 1]
```diff
@@ -287,12 +287,16 @@ mutate_cols <- function(.data, ...) {
 
       # only unchop if needed
       if (is.null(result)) {
-        result <- tryCatch(
-          vec_unchop(chunks, rows),
-          vctrs_error_incompatible_type = function(cnd) {
-            abort(class = "dplyr:::error_mutate_incompatible_combine", parent = cnd)
-          }
-        )
+        if (length(rows) == 1) {
```
Comment (romainfrancois) [inline R/mutate.R:290]: Not sure we can only base this on the number of groups, as it is perhaps possible that there is one group but `rows` not equal to `1:n()`

**Comment by romainfrancois** [comment]
Could we somehow work it out in `vec_unchop()`: 

``` r
library(dplyr, warn.conflicts = FALSE)

.Internal(inspect( group_rows(data.frame(x = c(1, 2))) ))
#> @7fd9bbd2d778 19 VECSXP g0c1 [OBJ,REF(2),ATT] (len=1, tl=0)
#>   @7fd9bbd38eb8 13 INTSXP g0c0 [REF(65535)]  1 : 2 (compact)
#> ATTRIB:
#>   @7fd9bbd42298 02 LISTSXP g0c0 [REF(1)] 
#>     TAG: @7fd9b724b5e8 01 SYMSXP g1c0 [MARK,REF(15)] "ptype"
#>     @7fd9bbd385f8 13 INTSXP g0c0 [REF(2)] (len=0, tl=0)
#>     TAG: @7fd9b400d1d0 01 SYMSXP g1c0 [MARK,REF(34498),LCK,gp=0x6000] "class" (has value)
#>     @7fd9bbd11818 16 STRSXP g0c3 [REF(2)] (len=3, tl=0)
#>       @7fd9b5e12b48 09 CHARSXP g1c2 [MARK,REF(150),gp=0x60] [ASCII] [cached] "vctrs_list_of"
#>       @7fd9b5e12a48 09 CHARSXP g1c2 [MARK,REF(557),gp=0x60] [ASCII] [cached] "vctrs_vctr"
#>       @7fd9b4026650 09 CHARSXP g1c1 [MARK,REF(1328),gp=0x61] [ASCII] [cached] "list"
.Internal(inspect( group_rows(data.frame(x = c(1, 2), y = c(3, 3)) %>% group_by(y)) ))
#> @7fd9bbfe24a0 19 VECSXP g0c1 [OBJ,REF(20),ATT] (len=1, tl=0)
#>   @7fd9bbfa0a38 13 INTSXP g0c1 [REF(9)] (len=2, tl=0) 1,2
#> ATTRIB:
#>   @7fd9bbfeb288 02 LISTSXP g0c0 [REF(1)] 
#>     TAG: @7fd9b724b5e8 01 SYMSXP g1c0 [MARK,REF(60)] "ptype"
#>     @7fd9bbf9bc38 13 INTSXP g0c0 [REF(7)] (len=0, tl=0)
#>     TAG: @7fd9b400d1d0 01 SYMSXP g1c0 [MARK,REF(65535),LCK,gp=0x6000] "class" (has value)
#>     @7fd9bbf97728 16 STRSXP g0c3 [REF(65535)] (len=3, tl=0)
#>       @7fd9b5e12b48 09 CHARSXP g1c2 [MARK,REF(154),gp=0x60] [ASCII] [cached] "vctrs_list_of"
#>       @7fd9b5e12a48 09 CHARSXP g1c2 [MARK,REF(560),gp=0x60] [ASCII] [cached] "vctrs_vctr"
#>       @7fd9b4026650 09 CHARSXP g1c1 [MARK,REF(1335),gp=0x61] [ASCII] [cached] "list"
```

<sup>Created on 2020-06-29 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

Seems like for ungrouped data frames, we get a compact representation of 1:n, but unfortunately not on the grouped with only one group case (it does appear unlikely that we see one such case)

**Comment by romainfrancois** [comment]
Perhaps something like this: 

```r
        skip_unchop <- TRUE
        if (is_grouped_df(.data) || inherits(.data, "rowwise_df")) {
          skip_unchop <- length(rows) == 1 && identical(rows[[1L]], seq_len(nrow(.data)))
        }
```

I'm just concerned about weird cases where there is only one group, but the rows are scrambled for some reason

**Comment by lionel-** [comment]
Maybe it should use `vec_slice()` in the 1 group case? @DavisVaughan Can we detect altrep sequences from base R reliably to make it a no-op in vec-slice?

**Comment by romainfrancois** [comment]
using `vec_slice()` however brings back the original problem: 

```
test-mutate.r:149: failure: unchop only called for when multiple groups
out$x is not an S3 object

test-mutate.r:149: failure: unchop only called for when multiple groups
out$x inherits from `integer` not `ts`.

test-mutate.r:153: failure: unchop only called for when multiple groups
out$x is not an S3 object

test-mutate.r:153: failure: unchop only called for when multiple groups
out$x inherits from `integer` not `ts`.
```

**Comment by romainfrancois** [comment]
Even with : 

```r
         if (is_grouped_df(.data) || inherits(.data, "rowwise_df")) {
            result <- vec_slice(chunks[[1]], rows[[1]])
          } else {
            result <- chunks[[1]]
          }
```

we still get: 

```
test-mutate.r:153: failure: unchop only called for when multiple groups
out$x is not an S3 object

test-mutate.r:153: failure: unchop only called for when multiple groups
out$x inherits from `integer` not `ts`.
```

from this test: 

```r
test_that("unchop only called for when multiple groups", {
  df <- data.frame(g = 1, x = 1:5)
  out <- mutate(df, x = ts(x, start = c(1971, 1), frequency = 52))
  expect_s3_class(out$x, "ts")

  gdf <- group_by(df, g)
  out <- mutate(gdf, x = ts(x, start = c(1971, 1), frequency = 52))
  expect_s3_class(out$x, "ts"). # <------- here
})
```

**Comment by DavisVaughan** [comment]
> Can we detect altrep sequences from base R reliably

I'm fairly certain you can't. `R_altrep_inherits(SEXP, R_altrep_class_t)` is exposed, but I'm pretty sure this is mainly for classes that you own. 


We'd need the `R_altrep_class_t` for `compact_intseq` objects to check against, which is created [here](https://github.com/wch/r-source/blob/2c72c07504d6d126b9f56f40535f250fe4221824/src/main/altclasses.c#L287), but isn't exposed in any way. I don't _think_ we can just call `R_make_altinteger_class()` again to get our own copy, because this messes with the registry of existing known altrep classes and their methods.

**Comment by romainfrancois** [comment]
I have mixed feelings about this, it is fairly rare to have a grouped data frame with only one group so we probably wouldn't pay too often for the comparison. I still think we should check it though.

**Comment by romainfrancois** [comment]
I'll let you decide @Hadley, I'm fine if we revert the last commit and forget about the problem it tries to solve, which isn't very likely to occur anyway, i.e. I don't think `dplyr` itself can make such 1 group with scrambled rows data frame, perhaps something like `strapgod` @DavisVaughan ? but then again `mutate()` does not make much sense there and I'd assume there would be more than one sample ...

**Comment by hadley** [comment]
Yeah, I think we should revert the last commit and merge as is. Do you mind doing that?

**Pull request merged**